src/Analyzer/Passes/LogicalExpressionOptimizerPass.cpp

```diff
@@ -173,6 +172,290 @@ bool isTwoArgumentsFromDifferentSides(const FunctionNode & node_function, const
            (first_src->isEqual(rhs_join) && second_src->isEqual(lhs_join));
 }
 
+void insertIfNotPresentInSet(QueryTreeNodePtrWithHashSet& set, QueryTreeNodes &nodes, QueryTreeNodePtr node)
+{
+    const auto [_, inserted] = set.emplace(node);
+    if (inserted)
+        nodes.push_back(std::move(node));
+}
+
+// Returns the flattened AND/OR node if the passed-in node can be flattened. Doesn't modify the passed-in node.
+std::shared_ptr<FunctionNode> getFlattenedAndOrOr(const FunctionNode & node, const ContextPtr & context)
+{
+    const auto & function_name = node.getFunctionName();
+    if (function_name != "or" && function_name != "and")
+        return nullptr;
+
+    const auto & arguments = node.getArguments().getNodes();
+    QueryTreeNodePtrWithHashMap<QueryTreeNodes> arguments_to_replace;
```

Comment:
You can replace it with `vector` of "flattened" arguments and a bool if any argument was flattened. It'll allow you to avoid lookups in the hash table.